### PR TITLE
Revive target: ESP8266 

### PR DESCRIPTION
**Pull request by rerobika**
This patch updates ESP8266 build system and code base as well. 
 - Removed unnecessary files.
 - Decreased code size.  
 - Refactored jerry_{port, extapi, run}.c to make it more easy to handle.
 - Readme.md is updated as well which contains detailed step by step description about how to set up environment and use JerryScript.      

Finally, it solves the related issue #1375.

JerryScript-DCO-1.0-Signed-off-by: Robert Fancsik frobert@inf.u-szeged.hu

### CODE REVIEW

**Comment by LaszloLango** [comment]
Thanks for this fix. Please mention the related issue (#1375) in the commit message.

File: targets/esp8266/user/jerry_run.c [hunk 1]
```diff
@@ -37,7 +37,7 @@ int js_entry (const char *source_p, const size_t source_size)
 
   js_register_functions ();
 
-  parsed_res = jerry_parse ((jerry_char_t *) jerry_src, source_size, false);
+  parsed_res = jerry_eval ((jerry_char_t *) jerry_src, source_size, false);
```
Comment (akosthekiss) [inline targets/esp8266/user/jerry_run.c]: Update error message 4 lines below. Or just drop the whole parse/eval functionality from this function and use this for jerry initialization only (and rewrite user_main.c to use `js_eval()` to run all scripts in `js_codes[]`).

And regarding initialization: don't forget to call `srand` as pointed out in #2056.

Comment (rerobika) [reply targets/esp8266/user/jerry_run.c]: I've added these modifies.

Comment (rerobika) [reply targets/esp8266/user/jerry_run.c]: I've found a bug that gettimeofday() always return with zero. I'm on to fix it.

File: targets/esp8266/Makefile [hunk 2]
```diff
@@ -73,32 +75,42 @@ LINKFLAGS_eagle.app.v6 =				\
...
+	-ljerry-core					\
+	-ljerry-libm					\
 	$(DEP_LIBS_eagle.app.v6)			\
-	./libs/lib_a-setjmp.o				\
 	-Wl,--end-group
 
 
 DEPENDS_eagle.app.v6 =					\
 	$(LD_FILE)					\
 	$(LDDIR)/eagle.rom.addr.v6.ld			\
```
Comment (akosthekiss) [inline targets/esp8266/Makefile]: are these two lines OK? the linker script is copied from the sdk and modified here (LDDIR is `$(SDK_PATH)/ld` if I'm not mistaken), and LD_FILE is defined only _after_ this line.

File: targets/esp8266/docs/ESP-PREREQUISITES.md [hunk 3]
```diff
@@ -30,7 +30,7 @@ sudo chown $USER /opt/Espressif/
 dependencies for x64
 ```
 sudo apt-get install git autoconf build-essential gperf \
-    bison flex texinfo libtool libncurses5-dev wget \
+    bison flex texinfo libtool libtool-bin libncurses5-dev wget \
```
Comment (akosthekiss) [inline targets/esp8266/docs/ESP-PREREQUISITES.md]: x86 and x64 deps are too repetitive. there should be no need for maintaining two almost identical copies of lists and steps. just present it once. and add the arch-dependent steps afterwards. (if I see it right, it's only the libc6-dev-i386 vs libc6-dev-amd64 package)

while ate it, fix the indentation of the line (this line is off by one space, next line is OK)

File: targets/esp8266/readme.md [hunk 4]
```diff
@@ -13,38 +13,18 @@ You can view online from
...
-
-Before flashing you need to follow the steps.
-
-1. Power off ESP8266
-2. Connect GPIO0 to GND and GPIO2 to VCC
-3. Power on ESP8266
-4. Flash
-
-##### 4.2 Flashing
+#### 3. Flashing for ESP8266 12E
```
Comment (akosthekiss) [inline targets/esp8266/readme.md]: the previous readme was for `ESP8266 ESP-01`, now this one states `ESP8266 12E`. is the previous board not supported anymore?

File: targets/esp8266/readme.md [hunk 5]
```diff
@@ -56,26 +36,37 @@ Default USB device is `/dev/ttyUSB0`. If you have different one, give with `USBD
...
-2. Disonnect(float) both GPIO0 and GPIO2
-3. Power on
+```
+make -f ./targets/esp8266/Makefile.esp8266 clean
+```
 
-Sample program here works with LED and a SW with below connection.
-* Connect GPIO2 to a LED > 4K resistor > GND
-* Connect GPIO0 between VCC > 4K resistor and GND
+If the board already been flashed it may contains written areas which will not be overwritten by flashing jerryscript.
```
Comment (akosthekiss) [inline targets/esp8266/readme.md]: this sentence needs a bit more massaging. "has already been flashed", "may contain": I'm almost sure that these changes are needed. but I'm not sure what the "written areas which will not be overwritten by flashing" means.

File: targets/esp8266/user/user_main.c [hunk 6]
```diff
@@ -43,20 +42,19 @@ void show_free_mem(int idx) {
...
 
+  js_entry();
+
   /* run main.js */
   show_free_mem(2);
-  retcode = js_entry(js_codes[0].source, js_codes[0].length);
+  retcode = js_eval(js_codes[0].source, js_codes[0].length);
   if (retcode != 0) {
-    printf("js_entry failed code(%d) [%s]\r\n", retcode, js_codes[0].name);
+    printf("js_eval failed code(%d) [%s]\r\n", retcode, js_codes[0].name);
```
Comment (akosthekiss) [inline targets/esp8266/user/user_main.c]: cannot this code simply be merged into the loop below? just start `src` from 0. oh, and yes: declare all the variables closest to their first use: `src` in the loop header, `retcode` in the loop body.

File: targets/esp8266/Makefile (hunk 2 above)
Comment (rerobika) [reply targets/esp8266/Makefile]: The second line is OK, the first line needs a bit modification as you pointed to this problem.

File: targets/esp8266/Makefile.esp8266 [hunk 7]
```diff
@@ -46,61 +38,38 @@ ifneq ($(MFORCE32),)
...
-
-	make -C $(INTERM) $(TYPE).external
-	cp `cat $(INTERM)/$(TYPE).external/list` $(OUTPUT)/.
-	cp $(OUTPUT)/lib$(TYPE).jerry-core.a $(COPYTARGET)/libjerrycore.a
-	cp $(OUTPUT)/lib$(TYPE).jerry-libm.lib.a $(COPYTARGET)/libjerrylibm.a
-	cp $(INTERM)/lib$(TYPE).external-entry.a $(OUTPUT)/.
-	cp $(OUTPUT)/lib$(TYPE).external-entry.a $(COPYTARGET)/libjerryentry.a
+	 -DMEM_HEAP_SIZE_KB=$(JERRYHEAP)
 
+	make -C$(BUILD_DIR) jerry-core jerry-port-default-minimal
```
[the remark this answers is not in the document]
Comment (rerobika) [reply targets/esp8266/Makefile.esp8266]: I forgot to remove it, it's unnecessary.

File: targets/esp8266/Makefile.esp8266 [hunk 8]
```diff
@@ -46,61 +38,38 @@ ifneq ($(MFORCE32),)
...
+	cp $(BUILD_DIR)/lib/libjerry-core.a $(COPYTARGET)/libjerry-core.a
```
[the remark this answers is not in the document]
Comment (rerobika) [reply targets/esp8266/Makefile.esp8266]: Good idea!

File: targets/esp8266/Makefile.esp8266 [hunk 9]
```diff
@@ -46,61 +38,38 @@ ifneq ($(MFORCE32),)
...
+	cp $(BUILD_DIR)/lib/libjerry-libm.a $(COPYTARGET)/libjerry-libm.a
```
[the remark this answers is not in the document]
Comment (rerobika) [reply targets/esp8266/Makefile.esp8266]: Can be removed it's unnecessary.

File: targets/esp8266/Makefile.esp8266 [hunk 10]
```diff
@@ -46,61 +38,38 @@ ifneq ($(MFORCE32),)
...
 
 js2c:
-	cd targets/esp8266; ../../tools/js2c.py
-
+	cd targets/esp8266; \
+	../../tools/js2c.py --dest ./include --js-source ./js
```
[the remark this answers is not in the document]
Comment (rerobika) [reply targets/esp8266/Makefile.esp8266]: It would.

File: targets/esp8266/Makefile.esp8266 [hunk 11]
```diff
@@ -46,61 +38,38 @@ ifneq ($(MFORCE32),)
...
 
 mkbin:
 	cd targets/esp8266; \
-	 make clean; \
-	 BOOT=new APP=1 SPI_SPEED=40 SPI_MODE=QIO SPI_SIZE_MAP=3 make; \
-	 BOOT=new APP=2 SPI_SPEED=40 SPI_MODE=QIO SPI_SIZE_MAP=3 make
-
+	make clean; \
+	make BOOT=new APP=0 SPI_SPEED=40 SPI_MODE=DIO SPI_SIZE_MAP=4
```
[the remark this answers is not in the document]
Comment (rerobika) [reply targets/esp8266/Makefile.esp8266]: Ditto.

File: targets/esp8266/docs/ESP-PREREQUISITES.md (hunk 3 above)
Comment (rerobika) [reply targets/esp8266/docs/ESP-PREREQUISITES.md]: I see your point, I'm going to reorder these deps.

File: targets/esp8266/readme.md (hunk 4 above)
Comment (rerobika) [reply targets/esp8266/readme.md]: Unfortunately it is not supported anymore. The toolchain generates two binaries:
- eagle.flash.bin ~50kB
- eagle.irom0text.bin ~ 400 kB

The ESP8266 ESP-01 has only 512 kB flash memory. Unluckily you cannot put these images next to each other (in this case it would fit into 512 kB flash size), because you cannot write your user data before 0x20000. It means you start flashing your user binary at around 131kB and 131 + 400 > 512.

File: targets/esp8266/readme.md (hunk 5 above)
Comment (rerobika) [reply targets/esp8266/readme.md]: I'll give more details what "written areas which will not be overwritten by flashing" means

File: targets/esp8266/user/user_main.c (hunk 6 above)
Comment (rerobika) [reply targets/esp8266/user/user_main.c]: Yeah, can be merged.

File: targets/esp8266/readme.md (hunk 5 above)
Comment (rerobika) [reply targets/esp8266/readme.md]: Let me correct myself, `erase_flash` rule is optional and doesn't need any prerequisites. So this description is unnecessary.

File: targets/esp8266/readme.md [hunk 12]
```diff
@@ -56,26 +36,36 @@ Default USB device is `/dev/ttyUSB0`. If you have different one, give with `USBD
...
-* Connect GPIO0 between VCC > 4K resistor and GND
+To clean the board's flash memory:
+```
+make -f ./targets/esp8266/Makefile.esp8266 erase_flash
+```
 
-If GPIO0 is High then LED is turned on longer. If L vice versa.
 
 ### 6. Optimizing initial RAM usage (ESP8266 specific)
 The existing open source gcc compiler with Xtensa support stores const(ants) in
```
Comment (akosthekiss) [inline targets/esp8266/readme.md]: "support stores" -> "supports storing"

File: targets/esp8266/docs/ESP-PREREQUISITES.md [hunk 13]
```diff
@@ -18,22 +18,24 @@ I've slightly changed the step to use SDK from Espressif official SDK
...
-    gawk libc6-dev-i386 python-serial libexpat-dev
+    bison flex texinfo libtool libtool-bin libncurses5-dev wget \
+    gawk python-serial libexpat-dev
 sudo mkdir /opt/Espressif
 sudo chown $USER /opt/Espressif/
+
 ```
 
-dependencies for x64
+x86 specified dependency:
```
Comment (akosthekiss) [inline targets/esp8266/docs/ESP-PREREQUISITES.md]: I'd use "specific" rather than "specified". E.g., "dependency specific to x64"

File: targets/esp8266/docs/ESP-PREREQUISITES.md [hunk 14]
```diff
@@ -18,22 +18,24 @@ I've slightly changed the step to use SDK from Espressif official SDK
...
 ```
-sudo apt-get install git autoconf build-essential gperf \
-    bison flex texinfo libtool libncurses5-dev wget \
-    gawk libc6-dev-amd64 python-serial libexpat-dev
-sudo mkdir /opt/Espressif
-sudo chown $USER /opt/Espressif/
+sudo apt-get install libc6-dev-i386
+```
+
+x64 specified dependency:
```
Comment (akosthekiss) [inline targets/esp8266/docs/ESP-PREREQUISITES.md]: ditto

File: targets/esp8266/Makefile.esp8266 [hunk 15]
```diff
@@ -46,61 +38,35 @@ ifneq ($(MFORCE32),)
...
 	 -DENABLE_LTO=OFF \
-	 -DENABLE_VALGRIND=OFF \
-	 -DCMAKE_TOOLCHAIN_FILE=build/configs/toolchain_external.cmake \
+	 -DFEATURE_VALGRIND=OFF \
+	 -DCMAKE_TOOLCHAIN_FILE=cmake/toolchain_external.cmake \
 	 -DEXTERNAL_CMAKE_SYSTEM_PROCESSOR=xtensia-lx106 \
 	 -DEXTERNAL_CMAKE_C_COMPILER=xtensa-lx106-elf-gcc \
 	 -DEXTERNAL_CMAKE_C_COMPILER_ID=GNU \
-	 -DEXTERNAL_BUILD_ENTRY_FILE="$(JERRY_BUILD_FILES)" \
+	 -DJERRY_LIBC=OFF \
```
Comment (akosthekiss) [inline targets/esp8266/Makefile.esp8266]: As jerry-libm is not needed, it should be disabled for cmake, too (`-DJERRY_LIBM=OFF`)

File: targets/esp8266/Makefile.esp8266 [hunk 16]
```diff
@@ -46,61 +38,35 @@ ifneq ($(MFORCE32),)
 	ESP_CFLAGS += -mforce-l32
```
Comment (akosthekiss) [inline targets/esp8266/Makefile.esp8266]: why not have this `MFORCE32`-dependent `ESP_CFLAGS` extension together with the extension above (where `JERRY_CONST_DATA` is defined)

File: targets/esp8266/Makefile [hunk 17]
```diff
@@ -52,53 +60,59 @@ TARGET_LDFLAGS =					\
...
-
+	user/libuser.a
 
 LINKFLAGS_eagle.app.v6 =				\
 	-L$(SDK_PATH)/lib				\
 	-Wl,--gc-sections				\
 	-Wl,-Map,output.map				\
 	-nostdlib					\
-	-T$(LD_FILE)					\
+	-T$(JERRY_LD_FILE)					\
```
Comment (akosthekiss) [inline targets/esp8266/Makefile]: style: trailing backslash wandered away, would be nice to fix it

File: targets/esp8266/Makefile [hunk 18]
```diff
@@ -52,53 +60,59 @@ TARGET_LDFLAGS =					\
...
-	-ljerrylibm					\
+	-ljerry-core					\
 	$(DEP_LIBS_eagle.app.v6)			\
-	./libs/lib_a-setjmp.o				\
 	-Wl,--end-group
 
-
 DEPENDS_eagle.app.v6 =					\
-	$(LD_FILE)					\
+	$(JERRY_LD_FILE)					\
```
Comment (akosthekiss) [inline targets/esp8266/Makefile]: ditto

**Comment by rerobika** [comment]
@akosthekiss Thanks for your review. During this refactoring I realized that jerry-limb is still needed, because SDK's libm is not proper. Including jerry-lib can solve it.

File: targets/esp8266/user/jerry_run.c [hunk 19]
```diff
@@ -27,25 +28,13 @@ jerry_value_t parsed_res;
...
-  const jerry_char_t *jerry_src = (const jerry_char_t *) source_p;
-  int ret_code = 0; /* JERRY_COMPLETION_CODE_OK */
+  srand ((unsigned) jerry_port_get_current_time());
   jerry_init_flag_t flags = JERRY_INIT_EMPTY;
 
   jerry_init (flags);
-
   js_register_functions ();
-
-  parsed_res = jerry_parse ((jerry_char_t *) jerry_src, source_size, false);
```
[the remark this answers is not in the document]
Comment (akosthekiss) [reply targets/esp8266/user/jerry_run.c]: Thanks for fixing up `js_eval`. It would be great if `js_loop` could be touched a bit, too. A lot of variable declarations could be (and should be, IMO) moved closer to their first use (and potentially merged with their initialization). (Especially `val_args` could be initialised in a simpler way: `jerry_value_t val_args[] = { jerry_create_number (ticknow) };`, and instead of the current `val_argv` we could have `sizeof (val_args) / sizeof (val_args[0])` or `sizeof (val_args) / sizeof (jerry_value_t)`, which would be more maintainable.)

File: targets/mbed/source/jerry_extapi.cpp [hunk 20]
```diff
@@ -27,12 +27,12 @@
 
 //-----------------------------------------------------------------------------
 
-#define __UNSED__ __attribute__((unused))
+#define __UNUSED__ __attribute__((unused))
```
Comment (akosthekiss) [inline targets/mbed/source/jerry_extapi.cpp]: I've just realized that this file is from the mbed target. This PR should not change other targets but ESP alone. Even if the mbed target contains questionable parts. Fixing those should be left for another PR.

File: targets/esp8266/user/jerry_extapi.c [hunk 21]
```diff
@@ -92,25 +84,29 @@ DELCARE_HANDLER(print) {
...
   }
   return jerry_create_boolean (true);
-}
-
-
-/*---------------------------------------------------------------------------*/
+} /* print */
 
 DELCARE_HANDLER(gpio_dir) {
   int port, value;
```
Comment (akosthekiss) [inline targets/esp8266/user/jerry_extapi.c]: Just about everywhere in this source: please move all variable declarations closest to their first use, preferably merging declaration and initialisation.

**Comment by akosthekiss** [comment]
A separate note: this PR is a lot more now than just updating the build system of the ESP target. This is actually reviving it, build system and code base alike. Please reflect this both in the commit message and in the title of the PR.

**Comment by akosthekiss** [comment]
One more note (actually a reminder): please also mention the related issue in the commit message, as @LaszloLango requested.

**Comment by rerobika** [comment]
@akosthekiss Thanks for your review. I applied your ideas!

**Comment by akosthekiss** [comment]
@rerobika Please double check. The commit still touches the mbed target.

**Comment by rerobika** [comment]
I've forgotten to revert it. Now it's done.

File: targets/esp8266/user/jerry_run.c [hunk 22]
```diff
@@ -94,23 +68,25 @@ int js_loop (uint32_t ticknow)
     return -2;
   }
 
-  val_argv = 1;
-  val_args[0] = jerry_create_number (ticknow);
```
Comment (akosthekiss) [inline targets/esp8266/user/jerry_run.c]: one too many empty line

File: targets/esp8266/user/jerry_run.c [hunk 23]
```diff
@@ -94,23 +68,25 @@ int js_loop (uint32_t ticknow)
...
 
-  res = jerry_call_function (sysloop_func,
+  jerry_value_t val_args[] = { jerry_create_number (ticknow) };
+  uint16_t val_argv = sizeof (val_args) / sizeof (jerry_value_t);
+
+  jerry_value_t res = jerry_call_function (sysloop_func,
                              global_obj_val,
```
Comment (akosthekiss) [inline targets/esp8266/user/jerry_run.c]: fix indentation of arguments

File: targets/esp8266/user/jerry_run.c [hunk 24]
```diff
@@ -94,23 +68,25 @@ int js_loop (uint32_t ticknow)
...
                              val_argv);
 
   if (jerry_value_has_error_flag (res)) {
-    ret_code = -3;
+    jerry_release_value (sysloop_func);
+    jerry_release_value (global_obj_val);
+    return -3;
   }
 
-  jerry_release_value (res);
```
Comment (akosthekiss) [inline targets/esp8266/user/jerry_run.c]: `res` is not released anymore (neither here, not in the error handling branch above)

File: targets/esp8266/readme.md [hunk 25]
```diff
@@ -56,26 +36,36 @@ Default USB device is `/dev/ttyUSB0`. If you have different one, give with `USBD
...
+```
+make -f ./targets/esp8266/Makefile.esp8266 erase_flash
+```
 
-If GPIO0 is High then LED is turned on longer. If L vice versa.
 
 ### 6. Optimizing initial RAM usage (ESP8266 specific)
-The existing open source gcc compiler with Xtensa support stores const(ants) in
-the same limited RAM where our code needs to run. 
+The existing open source gcc compiler with Xtensa supports storing const(ants) in
```
Comment (akosthekiss) [inline targets/esp8266/readme.md]: Oh, I've misread the sentence last time. It is OK as it was. "gcc compiler with Xtensa support stores const(ants) in the same limited RAM" is all right. Please, revert to that version. Sorry, my bad.

File: targets/esp8266/docs/ESP-PREREQUISITES.md [hunk 26]
```diff
@@ -56,14 +58,9 @@ PATH=$PWD/builds/xtensa-lx106-elf/bin:$PATH
...
 ln -s ESP8266_RTOS_SDK.git ESP8266_SDK
-cd ESP8266_SDK
-git checkout -b jerry a2b413ad2996450fe2f173b6afab243f6e1249aa
 ```
 
-We use SDK 1.2.0 version which has stdlib.h and others. Latest 1.3.0 version,
-as of writing this document, doesn't have it.
-(If anyone knows how to use latest version, please add an issue or send a PR.)
-
+We use latest SDK master.
```
Comment (LaszloLango) [inline targets/esp8266/docs/ESP-PREREQUISITES.md]: Is this really a good idea? I'd prefer a fix version.

Comment (rerobika) [reply targets/esp8266/docs/ESP-PREREQUISITES.md]: Using the latest master requires less maintenance than if we use a fix version. Also if you check https://github.com/espressif/ESP8266_RTOS_SDK/commits/master you can see improvements and fixes are common. That's why I'd prefer the latest master but @akosthekiss what's opinion about it?

Comment (LaszloLango) [reply targets/esp8266/docs/ESP-PREREQUISITES.md]: > Using the latest master requires less maintenance than if we use a fix version.

This is not true at all.

Comment (akosthekiss) [reply targets/esp8266/docs/ESP-PREREQUISITES.md]: I agree with @LaszloLango . That another repository can add not only fixes and improvements but incompatible changes, too. And then, all of a sudden, this code will not be working/compiling anymore.

In general, when working with external repos, it's good practice to state the last known working revision. Whether it is "formalized" with git submodules or simply documented in a readme/doc (like here should be) is secondary. The key point is that the maintainer should actively update that reference when it's been confirmed that the new version/revision works -- not reactively, when it turns out that something got broken.

Comment (rerobika) [reply targets/esp8266/docs/ESP-PREREQUISITES.md]: Seems true. I'll update the documentation.

**Comment by rerobika** [comment]
Version if fixed as you requested!

File: targets/esp8266/Makefile.esp8266 (hunk 15 above)
Comment (LaszloLango) [reply targets/esp8266/Makefile.esp8266]: This isn't fixed. Jerry libm is till on. Is jerry-libm needed or not?

File: targets/esp8266/readme.md (hunk 12 above)
Comment (LaszloLango) [reply targets/esp8266/readme.md]: Please fix

File: targets/esp8266/user/jerry_run.c [hunk 27]
```diff
@@ -94,23 +68,26 @@ int js_loop (uint32_t ticknow)
     return -2;
   }
 
-  val_argv = 1;
-  val_args[0] = jerry_create_number (ticknow);
+  jerry_value_t val_args[] = { jerry_create_number (ticknow) };
```
Comment (LaszloLango) [inline targets/esp8266/user/jerry_run.c]: I don't see where `val_args` is freed.

File: targets/esp8266/Makefile.esp8266 (hunk 15 above)
Comment (akosthekiss) [reply targets/esp8266/Makefile.esp8266]: Conclusion was that jerry-libm is still needed as toolchain's own libm is not featureful enough. Might be that this was discussed in another comment thread.

File: targets/esp8266/readme.md (hunk 12 above)
Comment (akosthekiss) [reply targets/esp8266/readme.md]: No need for that. Again, this was discussed before. I've made a bad comment. Text was changed first, then reverted per my request. For some reason, the original bad change request comment became visible again.

Comment (rerobika) [reply targets/esp8266/readme.md]: > Oh, I've misread the sentence last time. It is OK as it was. "gcc compiler with Xtensa support stores const(ants) in the same limited RAM" is all right. Please, revert to that version. Sorry, my bad.

This change is reverted.

File: targets/esp8266/user/jerry_run.c (hunk 27 above)
Comment (rerobika) [reply targets/esp8266/user/jerry_run.c]: Me neither, but your are right that should be freed.

Comment (rerobika) [reply targets/esp8266/user/jerry_run.c]: I fixed it. Thanks for your review!
